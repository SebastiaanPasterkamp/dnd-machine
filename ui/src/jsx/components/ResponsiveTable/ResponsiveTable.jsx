import React from 'react';
import PropTypes from 'prop-types';
import {
    map,
} from 'lodash/fp';

import utils from '../../utils.jsx';

import './sass/_responsive-table.scss';

const ResponsiveTable = function({
    items,
    headerComponent: HeaderComponent,
    rowComponent: RowComponent,
    footerComponent: FooterComponent,
    className,
}) {
    if (!items) {
        return null;
    }

    const style = utils.makeStyle({}, [
        'responsive-table nice-table condensed bordered responsive',
        className,
    ]);

    return (
        <table className={style}>
            {HeaderComponent ? (
                <thead>
                    <HeaderComponent />
                </thead>
            ) : null}
            <tbody>
                {map(
                    (item) => (
                        <RowComponent
                            key={item.id}
                            {...item}
                        />
                    )
                )(items)}
                {FooterComponent ? <FooterComponent /> : null}
            </tbody>
        </table>
    );
};

ResponsiveTable.propTypes = {
    items: PropTypes.oneOfType([
        PropTypes.arrayOf( PropTypes.shape({
            id: PropTypes.number.isRequired,
        }) ),
        PropTypes.objectOf( PropTypes.shape({
            id: PropTypes.number.isRequired,
        }) ),
    ]),
    headerComponent: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.func,
    ]),
    rowComponent: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.func,
    ]).isRequired,
    footerComponent: PropTypes.oneOfType([
        PropTypes.element,
        PropTypes.func,
    ]),
    className: PropTypes.string,
};


ResponsiveTable.defaultProps = {
    items: [],
    headerComponent: null,
    footerComponent: null,
    className: null,
};

export default ResponsiveTable;
