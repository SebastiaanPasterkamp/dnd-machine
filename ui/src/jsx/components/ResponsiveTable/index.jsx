export { default, ResponsiveTable } from './ResponsiveTable';
