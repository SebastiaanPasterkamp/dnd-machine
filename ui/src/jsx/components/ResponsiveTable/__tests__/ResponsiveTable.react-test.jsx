import React from 'react';
import { mount } from 'enzyme';
import ResponsiveTable from '../ResponsiveTable';
import renderer from 'react-test-renderer';

describe('Component: ResponsiveTable', () => {
    const fullProps = {
        items: [
            {id: 1, name: 'One'},
            {id: 2, name: 'Two'},
        ],
        headerComponent: () => (<tr><td>Name</td></tr>),
        rowComponent: (item) => (<tr><td id={item.id}>{item.name}</td></tr>),
        footerComponent: () => (<tr><td>Footer</td></tr>),
    };

    it('should render with minimum props', () => {
        const tree = renderer.create(
            <ResponsiveTable
                rowComponent={fullProps.rowComponent}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render with full props', () => {
        const tree = renderer.create(
            <ResponsiveTable
                {...fullProps}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

});
