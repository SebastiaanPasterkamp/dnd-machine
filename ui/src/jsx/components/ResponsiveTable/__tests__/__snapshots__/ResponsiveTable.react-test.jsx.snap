// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ResponsiveTable should render with full props 1`] = `
<table
  className="responsive-table nice-table condensed bordered responsive"
>
  <thead>
    <tr>
      <td>
        Name
      </td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        id={1}
      >
        One
      </td>
    </tr>
    <tr>
      <td
        id={2}
      >
        Two
      </td>
    </tr>
    <tr>
      <td>
        Footer
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Component: ResponsiveTable should render with minimum props 1`] = `
<table
  className="responsive-table nice-table condensed bordered responsive"
>
  <tbody />
</table>
`;
